lang/interpolate/parse.rl

```diff
@@ -0,0 +1,180 @@
+package interpolate
+
+import (
+	"fmt"
+)
+
+%%{
+	machine interpolate;
+	write data;
+}%%
+
+// Parse performs string interpolation on the input. It returns the list of
+// tokens found. It looks for variables of the format ${foo}. The curly braces
+// are required.
+func Parse(data string) (out Stream, _ error) {
+	var (
+		// variables used by Ragel
+		cs  = 0 // current state
+		p   = 0 // current position in data
+		pe  = len(data)
+		eof = pe // eof == pe if this is the last data block
+
+		// Index in data where the currently captured string started.
+		idx int
+
+		x string   // The string we use for holding a temporary value.
+		l Literal  // The string literal being read, if any.
+		v Variable // The variable being read, if any.
+
+		// Current token. This is either the variable that we just read
+		// or the string literal. We will append it to `out` and move
+		// on.
+		t Token
+	)
+
+	%%{
+		# Record the current position as the start of a string. This is
+		# usually used with the entry transition (>) to start capturing
+		# the string when a state machine is entered.
+		#
+		# fpc is the current position in the string (basically the same
+		# as the variable `p` but a special Ragel keyword) so after
+		# executing `start`, data[idx:fpc+1] is the string from when
+		# start was called to the current position (inclusive).
+		action start { idx = fpc }
+
+		# TODO: check this !
+		# A variable always starts with an alphabetical char and
+		# contains alphanumeric characters or numbers, underscores, and
+		# non-consecutive dots.
+		var_name
+			= ( [a-z] ([a-z0-9_]
+			  | ('.' | '_') [a-z0-9_])*
+			  )
+			>start
+			@{ v.Name = data[idx:fpc+1] }
+			;
+
+		# var is a reference to a variable.
+		var = '${' var_name '}'
+			;
+
+		#escape_sequence = ( '\\a'
+		#		  | '\\b'
+		#		  | '\\f'
+		#		  | '\\n'
+		#		  | '\\r'
+		#		  | '\\t'
+		#		  | '\\v'
+		#		  )
+		#	>start
+		#	@{ l = Literal{Value: "XXX" } }
+		#	;
+
+		# Anything followed by a '\' is matched and used appropriately.
+		#escaped_lit = '\\' any @{ l = Literal{Value: data[fpc:fpc+1]} };
+		#//l = Literal{Value: "Z("+data[fpc:fpc+1]+"):["+ data[fpc-1:fpc+1] +"]" }
+
+		# works!
```

Comment:
Why is this duplicated? It's almost the same as below? Is it here for documentation purposes? Should it be removed?

Reply:
This was my WIP branch that I pushed to get it out early. I've cleaned this up.